### PR TITLE
add evaluate code

### PR DESCRIPTION
**Pull request by fehiepsi**
In addressing #1, the R code is so slow and hard to understand (for me), so I need a review to double-check the math of my implementation. I'm pretty sure we can get expected weights for each time series from `M5Data.get_aggregated_ma_dollar_sales()` method. So I only need to review that given `weight`, `wrmsse` are `wspl` are computed as expected.

There are some issues which I will comment below.

### TODO

- [x] add a smoke test

### CODE REVIEW

File: evaluate.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    "rmse": eval_rmse,
+    "crps": eval_crps,
+    "pl": eval_pl,
+}
+
+
+def eval_wrmsse(raw_value, train_data, weight):
+    lag1 = train_data[..., 1:, :] - train_data[..., :-1, :]
+    # drop starting 0s
+    actual_length = (train_data.cumsum(-2) != 0).sum(-2)
```
Comment (fehiepsi) [inline evaluate.py]: This is an issue. The scaling value is calculated based on train_data but stripping the leading 0.

> Note that the denominator of RMSSE is computed only for the time-periods for which the examined product(s) are actively sold, i.e., the periods following the first non-zero demand observed for the series under evaluation.

We need to make sure that the smallest train_window has `actual_length >= 2`. I am not sure if we do it outside of backtesting and add a warning if we detect a violation. Or we just keep it as is.

Another issue is if the sale values are constant (unlikely to happen) during the train window, the `scale` (which is computed from `lag1`) can be 0. Hopefully, we won't face that issue.

Anyway, if we observe `inf` in `wrmsse` or `wspl`, one of the above two situations have been happening.

File: evaluate.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    Like pyro.contrib.forecast.eval_mae but does not average over batch dimensions.
+    """
+    pred = pred.median(0).values
+    return (pred - truth).abs().reshape(truth.shape[:-2] + (-1,)).mean(-1).cpu()
+
+
+@torch.no_grad()
+def eval_rmse(pred, truth):
+    """
+    Like pyro.contrib.forecast.eval_rmae but does not average over batch dimensions.
```
Comment (fehiepsi) [inline evaluate.py]: we have weight and scale factors for each time series (among a batch), so we need to keep batch dimension for the result.

File: evaluate.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    """
+    if weight is None:
+        weight = data.new_ones(data.shape[:-1])
+    assert weight.shape == data.shape[:-1]
+    weight = weight / weight.sum()
+
+    if kwargs.get("metrics") is None:
+        kwargs["metrics"] = DEFAULT_METRICS
+
+    windows = pyro_backtest(data, covariates, model_fn, **kwargs)
```
Comment (fehiepsi) [inline evaluate.py]: We won't log `eval_foo` result if we want to maintain the batch dimension. This should be fixed in Pyro backtest.

File: evaluate.py (hunk 1 above)
Comment (fritzo) [reply evaluate.py]: Can we simply specify `min_train_window` to `backtest()`? http://docs.pyro.ai/en/stable/contrib.forecast.html#pyro.contrib.forecast.evaluate.backtest

File: evaluate.py (hunk 2 above)
Comment (fritzo) [reply evaluate.py]: nit: s/eval_rmae/eval_rmse in the docstring

File: evaluate.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+
+def backtest(data, covariates, model_fn, weight=None, **kwargs):
+    """
+    Backtest function with weighted metrics. See
+    http://docs.pyro.ai/en/stable/contrib.forecast.html#pyro.contrib.forecast.evaluate.backtest
+    for more information.
+
+    .. note:: In M5 competition, joint result for all aggregation levels is
+        the average of the results at each aggregation level.
```
Comment (fehiepsi) [inline evaluate.py:85]: @fritzo i.e. the result of the single aggregated time-series has the same weight as the weighted result of 30k non-aggregated time series. This explains the top-down approach results of benchmark models.

File: evaluate.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    :param torch.Tensor weight: weight of each time series in `data`.
+        This should satisfy `weight.shape == data.shape[:-1]`.
+    """
+    if weight is None:
+        weight = data.new_ones(data.shape[:-1])
+    assert weight.shape == data.shape[:-1]
+    weight = weight / weight.sum()
```
Comment (fehiepsi) [inline evaluate.py]: bug: only normalize over batch dimensions.

File: evaluate.py (hunk 3 above)
Comment (fritzo) [reply evaluate.py]: Good point, I've added this one-line fix to https://github.com/pyro-ppl/pyro/pull/2364

File: evaluate.py (hunk 1 above)
Comment (fehiepsi) [reply evaluate.py]: Makes sense! I'll add a check for it.

File: evaluate.py (hunk 3 above)
Comment (fehiepsi) [reply evaluate.py]: thanks for the fix!

File: evaluate.py (hunk 2 above)
Comment (fehiepsi) [reply evaluate.py]: @fritzo I am not sure what to do with the batch result of `mae`, `cprs`, so I'll use the Pyro version here instead.

File: evaluate.py (hunk 4 above)
Comment (fritzo) [reply evaluate.py:85]: Well that seems like a justification for inconsistent models in my opinion. I think the Bayesian approach is to create an interpretable top-down generative model of latent parameters, then aggregate bottom-up from observations to aggregates. As I see it this is a basic methodological issue: reconciliation vs latent variable models. 🤔

File: evaluate.py (hunk 2 above)
Comment (fritzo) [reply evaluate.py]: We should be able to aggregate `mae`, `crps` and `rmse` in exactly the same way, since they all are in units of data (# items sold). Can you explain your concern?

File: evaluate.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    return (error * error).reshape(truth.shape[:-2] + (-1,)).mean(-1).sqrt()
+
+
+@torch.no_grad()
+def eval_pl(pred, truth):
+    """
+    Computes pinball loss over 9 quantiles 0.005, 0.165, 0.25, 0.5, 0.75, 0.835, 0.975, 0.995.
+    """
+    us = torch.tensor(M5Data.quantiles, dtype=pred.dtype, device=pred.device)
+    pred = quantile(pred, probs=us, dim=0)  # 9 x batch_shape x duration x D
```
Comment (fritzo) [inline evaluate.py:45]: Subtle point: I believe the `quantile` function is actually biased and we should fix it in Pyro. That's a separate discussion; I'll open a PR for discussion.

File: evaluate.py (hunk 2 above)
Comment (fehiepsi) [reply evaluate.py]: So I believe for `mae` and `crps`, we will compute `scale` using `lag1_l1` norm instead of `l2` norm. Will do so.

File: evaluate.py (hunk 4 above)
Comment (fehiepsi) [reply evaluate.py:85]: Well, I get your point. And I am not sure how to use the top-down approach for uncertainty.

**Comment by fehiepsi** [comment]
@fritzo Things look good to me. I'm going to construct a model now.

File: evaluate.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+def _get_metric_scale(metric, train_data):
+    duration = train_data.shape[-2]
+    lag1 = train_data - torch.nn.functional.pad(train_data[..., :-1, :], (0, 0, 1, 0))
+    # find active time: to drop the leading 0s
+    active_time = (train_data.sum(-1, keepdims=True).cumsum(-2) != 0).sum(-2, keepdims=True)
+    start_value = train_data.gather(
+        -2, (duration - active_time).expand(active_time.shape[:-1] + train_data.shape[-1:]))
+    norm = 2 if metric == "rmse" else 1
+    lag1_norm = lag1.abs().pow(norm).sum(-2) - start_value.squeeze(-2).abs().pow(norm)
```
Comment (fritzo) [inline evaluate.py:68]: I don't quite understand the `start_value` bit; could it be avoided if you instead defined
```py
lag1 = tran_data[..., 1:, :] - train_data[..., :-1, :]
```

Comment (fehiepsi) [reply evaluate.py:68]: This is a bit subtle: assume x = train_data = (0, 0, 1, 3, 6); the active series will be (1, 3, 6) and active lag1 is (2, 3). If we use lag1 = x[1:] - x[:-1], then lag1 = (0, 1, 2, 3). So we need to remove `1`, which is the `start_value`, to get active lag1.

File: evaluate.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+@torch.no_grad()
+def eval_pl(pred, truth):
+    """
+    Computes pinball loss over 9 quantiles 0.005, 0.165, 0.25, 0.5, 0.75, 0.835, 0.975, 0.995.
+    """
+    us = torch.tensor(M5Data.quantiles, dtype=pred.dtype, device=pred.device)
+    pred = quantile(pred, probs=us, dim=0)  # 9 x batch_shape x duration x D
+    error = pred - truth.unsqueeze(0)
+    us = us.reshape((-1,) + (1,) * (pred.dim() - 1)).expand(pred.shape)
+    error = (torch.where(error <= 0, us, 1 - us) * error).mean(0)  # mean accross all quantiles
```
Comment (fritzo) [inline evaluate.py]: I think you're missing an `.abs()`, or you could simplify to:
```py
error = torch.where(error <= 0, -us, 1 - us).mul(error).mean(0)
```

**Review by fritzo — APPROVED** [review]
Well I don't see any more errors, but I can't say I understand the R code linked. I think this is probably correct.
